extensions/standard-processors/processors/RouteText.cpp

```diff
@@ -0,0 +1,474 @@
+
+ public:
+  MatchingContext(core::ProcessContext& process_context, std::shared_ptr<core::FlowFile> flow_file, CasePolicy case_policy)
+    : process_context_(process_context),
+      flow_file_(std::move(flow_file)),
+      case_policy_(case_policy) {}
+
+  const std::regex& getRegexProperty(const core::Property& prop) {
+    auto it = regex_values_.find(prop.getName());
+    if (it != regex_values_.end()) {
+      return it->second;
+    }
+    std::string value;
+    if (!process_context_.getDynamicProperty(prop, value, flow_file_)) {
+      throw Exception(PROCESSOR_EXCEPTION, "Missing dynamic property: '" + prop.getName() + "'");
+    }
+    std::regex::flag_type flags = std::regex::ECMAScript;
+    if (case_policy_ == CasePolicy::IGNORE_CASE) {
+      flags |= std::regex::icase;
+    }
+    return (regex_values_[prop.getName()] = std::regex(value, flags));
+  }
+
+  const std::string& getStringProperty(const core::Property& prop) {
+    auto it = string_values_.find(prop.getName());
+    if (it != string_values_.end()) {
+      return it->second;
+    }
+    std::string value;
+    if (!process_context_.getDynamicProperty(prop, value, flow_file_)) {
+      throw Exception(PROCESSOR_EXCEPTION, "Missing dynamic property: '" + prop.getName() + "'");
+    }
+    return (string_values_[prop.getName()] = value);
+  }
+
+  const Searcher& getSearcher(const core::Property& prop) {
+    auto it = searcher_values_.find(prop.getName());
+    if (it != searcher_values_.end()) {
+      return it->second.searcher_;
+    }
+    std::string value;
+    if (!process_context_.getDynamicProperty(prop, value, flow_file_)) {
+      throw Exception(PROCESSOR_EXCEPTION, "Missing dynamic property: '" + prop.getName() + "'");
+    }
+
+    return searcher_values_.emplace(
+        std::piecewise_construct, std::forward_as_tuple(prop.getName()),
+        std::forward_as_tuple(value, case_policy_)).first->second.searcher_;
+  }
+
+  core::ProcessContext& process_context_;
+  std::shared_ptr<core::FlowFile> flow_file_;
+  CasePolicy case_policy_;
+
+  std::map<std::string, std::string> string_values_;
+  std::map<std::string, std::regex> regex_values_;
+
+  struct OwningSearcher {
+    OwningSearcher(std::string str, CasePolicy case_policy)
+      : str_(std::move(str)), searcher_(str_.cbegin(), str_.cend(), CaseAwareHash{case_policy}, CaseAwareEq{case_policy}) {}
+    OwningSearcher(const OwningSearcher&) = delete;
+    OwningSearcher(OwningSearcher&&) = delete;
+    OwningSearcher& operator=(const OwningSearcher&) = delete;
+    OwningSearcher& operator=(OwningSearcher&&) = delete;
+
+    std::string str_;
+    Searcher searcher_;
+  };
+
+  std::map<std::string, OwningSearcher> searcher_values_;
+};
+
+namespace {
+struct Route {
+  core::Relationship relationship_;
+  std::optional<std::string> group_name_;
+
+  bool operator<(const Route& other) const {
+    return std::tie(relationship_, group_name_) < std::tie(other.relationship_, other.group_name_);
+  }
```

Comment:
It looks like I was wrong here. The CI run failed because of this line. Not exactly sure why, let's just revert it.

Reply:
reverted, what could have been the problem?

Reply:
It looks like it was because of missing spaceship operator of Relationship. https://github.com/apache/nifi-minifi-cpp/runs/4077160988?check_suite_focus=true#step:4:10677
I expected it to be automatically available though operator rewrite rules, but apparently it's not.